<launch>
	<!-- option of camera pose estimation -->
	<arg name="markerNum" default="4"/>
	<arg name="sampleNum" default="500"/>
	<arg name="samplingTime" default="0.1"/>
	<arg name="image_enable" default="false"/>
	<arg name="pcd_enable" default="true"/>
	<arg name="distanceThreshold" default="0.4"/>

	<!-- position of ar_marker -->
	<arg name="cam00_enable" default="true"/>
	<arg name="cam01_enable" default="true"/>
	<arg name="cam02_enable" default="true"/>
	<arg name="cam03_enable" default="true"/>
	<arg name="cam04_enable" default="false"/>
	<arg name="cam05_enable" default="false"/>
	<arg name="cam06_enable" default="false"/>
	<arg name="cam07_enable" default="false"/>

	<!-- position of ar_marker -->
	<arg name="marker0_pos_x" default="0.695"/>
	<arg name="marker0_pos_y" default="-0.305"/>
	<arg name="marker0_pos_z" default="0.70"/>

	<arg name="marker1_pos_x" default="0.695"/>
	<arg name="marker1_pos_y" default="0.305"/>
	<arg name="marker1_pos_z" default="0.70"/>
	
	<arg name="marker2_pos_x" default="1.305"/>
	<arg name="marker2_pos_y" default="-0.305"/>
	<arg name="marker2_pos_z" default="0.70"/>

	<arg name="marker3_pos_x" default="1.305"/>
	<arg name="marker3_pos_y" default="0.305"/>
	<arg name="marker3_pos_z" default="0.70"/>
	
	<arg name="marker4_pos_x" default="4.00"/>
	<arg name="marker4_pos_y" default="-0.50"/>
	<arg name="marker4_pos_z" default="0.70"/>

	<arg name="marker5_pos_x" default="5.00"/>
	<arg name="marker5_pos_y" default="0.50"/>
	<arg name="marker5_pos_z" default="0.70"/>
	
	<arg name="marker6_pos_x" default="5.00"/>
	<arg name="marker6_pos_y" default="-0.50"/>
	<arg name="marker6_pos_z" default="0.70"/>
	
	<arg name="marker7_pos_x" default="5.00"/>
	<arg name="marker7_pos_y" default="0.50"/>
	<arg name="marker7_pos_z" default="0.70"/>
	

	<!-- option of AR marker tracker -->
	<arg name="marker_size" default="26.7" />
	<arg name="max_new_marker_error" default="0.001" />
	<arg name="max_track_error" default="0.002" />

	<!-- option of camera topic -->
	<arg name="cam_name" default="env_cam"/>
	<arg name="image_topic" default="/hd/color/image_rect_raw" />
	<arg name="info_topic" default="/hd/camera_info"/>	
	<arg name="pcd_topic" default="/hd/points" />
	<arg name="pcd_info_topic" default="/hd/camera_info"/>	
	<arg name="output_frame" default="_link" />

	<!-- ar pose server -->
	<node pkg="camerapose_server" name="arpose_server"  type="arpose_server" output="screen">
		<param name="markerNum" value="$(arg markerNum)"/>

		<param name="marker0_pos_x" value="$(arg marker0_pos_x)"/>
		<param name="marker0_pos_y" value="$(arg marker0_pos_y)"/>
		<param name="marker0_pos_z" value="$(arg marker0_pos_z)"/>

		<param name="marker1_pos_x" value="$(arg marker1_pos_x)"/>
		<param name="marker1_pos_y" value="$(arg marker1_pos_y)"/>
		<param name="marker1_pos_z" value="$(arg marker1_pos_z)"/>

		<param name="marker2_pos_x" value="$(arg marker2_pos_x)"/>
		<param name="marker2_pos_y" value="$(arg marker2_pos_y)"/>
		<param name="marker2_pos_z" value="$(arg marker2_pos_z)"/>

		<param name="marker3_pos_x" value="$(arg marker3_pos_x)"/>
		<param name="marker3_pos_y" value="$(arg marker3_pos_y)"/>
		<param name="marker3_pos_z" value="$(arg marker3_pos_z)"/>

		<param name="marker4_pos_x" value="$(arg marker4_pos_x)"/>
		<param name="marker4_pos_y" value="$(arg marker4_pos_y)"/>
		<param name="marker4_pos_z" value="$(arg marker4_pos_z)"/>

		<param name="marker5_pos_x" value="$(arg marker5_pos_x)"/>
		<param name="marker5_pos_y" value="$(arg marker5_pos_y)"/>
		<param name="marker5_pos_z" value="$(arg marker5_pos_z)"/>

		<param name="marker6_pos_x" value="$(arg marker6_pos_x)"/>
		<param name="marker6_pos_y" value="$(arg marker6_pos_y)"/>
		<param name="marker6_pos_z" value="$(arg marker6_pos_z)"/>

		<param name="marker7_pos_x" value="$(arg marker7_pos_x)"/>
		<param name="marker7_pos_y" value="$(arg marker7_pos_y)"/>
		<param name="marker7_pos_z" value="$(arg marker7_pos_z)"/>

	</node>

	<!-- camera pose saver -->
	<node pkg="camerapose_server" name="camerapose_saver"  type="camerapose_saver" output="screen">
		<param name="cam00_enable" value="$(arg cam00_enable)"/>
		<param name="cam01_enable" value="$(arg cam01_enable)"/>
		<param name="cam02_enable" value="$(arg cam02_enable)"/>
		<param name="cam03_enable" value="$(arg cam03_enable)"/>
		<param name="cam04_enable" value="$(arg cam04_enable)"/>
		<param name="cam05_enable" value="$(arg cam05_enable)"/>
		<param name="cam06_enable" value="$(arg cam06_enable)"/>
		<param name="cam07_enable" value="$(arg cam07_enable)"/>
		<param name="sampleNum" value="$(arg sampleNum)"/>
		<param name="output_frame" value="$(arg output_frame)"/>
		<param name="distanceThreshold" value="$(arg distanceThreshold)"/>
	</node>

	<!-- rviz -->
	<node pkg="rviz" type="rviz" name="ar_marker_rviz" args="-d $(find camerapose_server)/rviz/ar_marker_double.rviz" />

	<!-- cam00 -->
	<group if="$(arg cam00_enable)">
		<!-- arg_setting -->
		<arg name="cam00_id" default="00"/>
		<arg name="cam00_name" default="$(arg cam_name)$(arg cam00_id)"/>
		<arg name="cam00_image_topic" default="$(arg cam00_name)$(arg image_topic)" />
		<arg name="cam00_info_topic" default="$(arg cam00_name)$(arg info_topic)"/>	
		<arg name="cam00_pcd_topic" default="$(arg cam00_name)$(arg pcd_topic)" />
		<arg name="cam00_pcd_info_topic" default="$(arg cam00_name)$(arg pcd_info_topic)"/>	
		<arg name="output00_frame" default="$(arg cam00_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<!--node name="ar_track_alvar00" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam00_image_topic) $(arg cam00_info_topic) $(arg output00_frame) $(arg cam00_name)" /-->
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar00_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam00_pcd_topic) $(arg cam00_pcd_info_topic) $(arg output00_frame) $(arg cam00_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater00" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam00_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output00_frame)"/>
		</node>
	</group>
	
	<!-- cam01 -->
	<group if="$(arg cam01_enable)">
		<!-- arg_setting -->
		<arg name="cam01_id" default="01"/>
		<arg name="cam01_name" default="$(arg cam_name)$(arg cam01_id)"/>
		<arg name="cam01_image_topic" default="$(arg cam01_name)$(arg image_topic)" />
		<arg name="cam01_info_topic" default="$(arg cam01_name)$(arg info_topic)"/>	
		<arg name="cam01_pcd_topic" default="$(arg cam01_name)$(arg pcd_topic)" />
		<arg name="cam01_pcd_info_topic" default="$(arg cam01_name)$(arg pcd_info_topic)"/>	
		<arg name="output01_frame" default="$(arg cam01_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<!--node name="ar_track_alvar01" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam01_image_topic) $(arg cam01_info_topic) $(arg output01_frame) $(arg cam01_name)" /-->
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar01_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam01_pcd_topic) $(arg cam01_pcd_info_topic) $(arg output01_frame) $(arg cam01_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater01" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam01_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output01_frame)"/>
		</node>
	</group>

    <!-- cam02 -->
	<group if="$(arg cam02_enable)">
		<!-- arg_setting -->
		<arg name="cam02_id" default="02"/>
		<arg name="cam02_name" default="$(arg cam_name)$(arg cam02_id)"/>
		<arg name="cam02_image_topic" default="$(arg cam02_name)$(arg image_topic)" />
		<arg name="cam02_info_topic" default="$(arg cam02_name)$(arg info_topic)"/>	
		<arg name="cam02_pcd_topic" default="$(arg cam02_name)$(arg pcd_topic)" />
		<arg name="cam02_pcd_info_topic" default="$(arg cam02_name)$(arg pcd_info_topic)"/>	
		<arg name="output02_frame" default="$(arg cam02_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<!--node name="ar_track_alvar02" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam02_image_topic) $(arg cam02_info_topic) $(arg output02_frame) $(arg cam02_name)" /-->
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar02_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam02_pcd_topic) $(arg cam02_pcd_info_topic) $(arg output02_frame) $(arg cam02_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater02" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam02_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output02_frame)"/>
		</node>
	</group>

    <!-- cam03 -->
	<group if="$(arg cam03_enable)">
		<!-- arg_setting -->
		<arg name="cam03_id" default="03"/>
		<arg name="cam03_name" default="$(arg cam_name)$(arg cam03_id)"/>
		<arg name="cam03_image_topic" default="$(arg cam03_name)$(arg image_topic)" />
		<arg name="cam03_info_topic" default="$(arg cam03_name)$(arg info_topic)"/>	
		<arg name="cam03_pcd_topic" default="$(arg cam03_name)$(arg pcd_topic)" />
		<arg name="cam03_pcd_info_topic" default="$(arg cam03_name)$(arg pcd_info_topic)"/>	
		<arg name="output03_frame" default="$(arg cam03_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<!--node name="ar_track_alvar03" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam03_image_topic) $(arg cam03_info_topic) $(arg output03_frame) $(arg cam03_name)" /-->
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar03_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam03_pcd_topic) $(arg cam03_pcd_info_topic) $(arg output03_frame) $(arg cam03_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater03" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam03_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output03_frame)"/>
		</node>
	</group>

    <!-- cam04 -->
	<group if="$(arg cam04_enable)">
		<!-- arg_setting -->
		<arg name="cam04_id" default="04"/>
		<arg name="cam04_name" default="$(arg cam_name)$(arg cam04_id)"/>
		<arg name="cam04_image_topic" default="$(arg cam04_name)$(arg image_topic)" />
		<arg name="cam04_info_topic" default="$(arg cam04_name)$(arg info_topic)"/>	
		<arg name="cam04_pcd_topic" default="$(arg cam04_name)$(arg pcd_topic)" />
		<arg name="cam04_pcd_info_topic" default="$(arg cam04_name)$(arg pcd_info_topic)"/>	
		<arg name="output04_frame" default="$(arg cam04_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<!--node name="ar_track_alvar04" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam04_image_topic) $(arg cam04_info_topic) $(arg output04_frame) $(arg cam04_name)" /-->
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar04_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam04_pcd_topic) $(arg cam04_pcd_info_topic) $(arg output04_frame) $(arg cam04_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater04" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam04_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output04_frame)"/>
		</node>
	</group>
    
    <!-- cam05 -->
	<group if="$(arg cam05_enable)">
		<!-- arg_setting -->
		<arg name="cam05_id" default="05"/>
		<arg name="cam05_name" default="$(arg cam_name)$(arg cam05_id)"/>
		<arg name="cam05_image_topic" default="$(arg cam05_name)$(arg image_topic)" />
		<arg name="cam05_info_topic" default="$(arg cam05_name)$(arg info_topic)"/>	
		<arg name="cam05_pcd_topic" default="$(arg cam05_name)$(arg pcd_topic)" />
		<arg name="cam05_pcd_info_topic" default="$(arg cam05_name)$(arg pcd_info_topic)"/>	
		<arg name="output05_frame" default="$(arg cam05_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<node name="ar_track_alvar05" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam05_image_topic) $(arg cam05_info_topic) $(arg output05_frame) $(arg cam05_name)" />
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar05_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam05_pcd_topic) $(arg cam05_pcd_info_topic) $(arg output05_frame) $(arg cam05_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater05" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam05_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output05_frame)"/>
		</node>
	</group>

    <!-- cam06 -->
	<group if="$(arg cam06_enable)">
		<!-- arg_setting -->
		<arg name="cam06_id" default="06"/>
		<arg name="cam06_name" default="$(arg cam_name)$(arg cam06_id)"/>
		<arg name="cam06_image_topic" default="$(arg cam06_name)$(arg image_topic)" />
		<arg name="cam06_info_topic" default="$(arg cam06_name)$(arg info_topic)"/>	
		<arg name="cam06_pcd_topic" default="$(arg cam06_name)$(arg pcd_topic)" />
		<arg name="cam06_pcd_info_topic" default="$(arg cam06_name)$(arg pcd_info_topic)"/>	
		<arg name="output06_frame" default="$(arg cam06_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<node name="ar_track_alvar06" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam06_image_topic) $(arg cam06_info_topic) $(arg output06_frame) $(arg cam06_name)" />
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar06_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam06_pcd_topic) $(arg cam06_pcd_info_topic) $(arg output06_frame) $(arg cam06_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater06" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam06_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output06_frame)"/>
		</node>
	</group>

    <!-- cam07 -->
	<group if="$(arg cam07_enable)">
		<!-- arg_setting -->
		<arg name="cam07_id" default="07"/>
		<arg name="cam07_name" default="$(arg cam_name)$(arg cam07_id)"/>
		<arg name="cam07_image_topic" default="$(arg cam07_name)$(arg image_topic)" />
		<arg name="cam07_info_topic" default="$(arg cam07_name)$(arg info_topic)"/>	
		<arg name="cam07_pcd_topic" default="$(arg cam07_name)$(arg pcd_topic)" />
		<arg name="cam07_pcd_info_topic" default="$(arg cam07_name)$(arg pcd_info_topic)"/>	
		<arg name="output07_frame" default="$(arg cam07_name)$(arg output_frame)" />

		<!-- ar_track_alvar image -->
		<node name="ar_track_alvar07" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam07_image_topic) $(arg cam07_info_topic) $(arg output07_frame) $(arg cam07_name)" />
		<!-- ar_track_alvar pcd -->
		<node name="ar_track_alvar07_pcd" pkg="ar_track_alvar" type="individualMarkers" respawn="false" output="screen"
		args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam07_pcd_topic) $(arg cam07_pcd_info_topic) $(arg output07_frame) $(arg cam07_name)" />
		<!-- camerapose_estimater -->
		<node name="camerapose_estimater07" pkg="camerapose_server" type="camerapose_estimater" output="screen">
			<param name="cameraName" value="$(arg cam07_name)"/>
			<param name="markerNum" value="$(arg markerNum)"/>
			<param name="pcd_enable" value="$(arg pcd_enable)"/>
			<param name="image_enable" value="$(arg image_enable)"/>
			<param name="output_frame" value="$(arg output07_frame)"/>
		</node>
	</group>
	
</launch>